### ISSUE TITLE
Uniform buffer block size issue on Metal

### ISSUE THREAD

**Issue by haoxli** [gpuweb/gpuweb#425]
On Metal, it is required that the size of the constant buffer must be no less than the size of the constant buffer structure defined in shader, and the overall size of the structure declared in the constant buffer is aligned to the largest base alignment of its members.

For example, considering the following constant buffer declaration,
```
struct Block {
    float4 color;
    float depth;
};

fragment FragOut main(constant Block& uBuffer [[buffer(0)]]) {
    ...
}
```
As the largest alignment of the attribute (float4) is 16 bytes,  the actual size of `uBuffer` is 32 bytes, so when we bind a buffer with 20 bytes in bind group, the Metal validation layer will report the following validation error:
> Fragment Function(main0): argument uBuffer[0] from buffer(0) with offset(0) and length(20) has space for 20 bytes, but argument has a length(32)

For constant buffer declaration below, as the largest alignment of the attribute of `uBuffer` is 16 bytes,  the size of the constant buffer bound to `[[buffer(0)]]` must be at least 48 bytes.
```
struct Block {
    float2 a;
    float3 b;
    float4 c;
};

fragment FragOut main(constant Block& uBuffer [[buffer(0)]]) {
    ...
}
```

There is a similar issue on gfx-rs: https://github.com/gfx-rs/wgpu/issues/185

In conclusion, in WebGPU we should validate the size of the uniform buffer bound in bind group must be no less than the size of the uniform block declared in shader, and the computation of the uniform block size must be compatible with the algorithm in Metal.

**Comment by Kangz** [comment]
Another solution that was brought up in Dawn was that we could internally align the size of buffers to the maximum alignment of Metal structure and ideally with all the out of bounds checks the applications wouldn't be able to tell.

**Comment by kvark** [comment]
@dneto0 could we get some clarification on this vs #1558 ?
Consensus here: this uniform structure needs to have 32 bytes
Consensus in #1558: this uniform structure can be 20 bytes

**Comment by Kangz** [comment]
I think the spec is clear on this now.

WebGPU spec the buffer range bound should be at least `SizeOf(Block)` bytes in size:

> If the corresponding shader structure doesn’t end with an unbounded array type, then the value of entry.buffer.minBindingSize must be greater than or equal to the size of the structure.

WGSL spec for `SizeOf(struct Block {float4 color; float depth;};)`

> AlignOf(color) = AlignOf(vec4<f32>) = 16
> SizeOf(color) = SizeOf(vec4<f32>) = 16
> AlignOf(depth) = AlignOf(f32) = 4
> SizeOf(depth) = SizeOf(f32) = 4
> AlignOf(Block) = max(AlignOfMember(Block, color), AlignOfMember(Block, depth)) = max(16, 4) = 16
> sizeof(Block) =  roundUp(AlignOf(Block), OffsetOfMember(Block, depth) + SizeOfMember(Block, depth)) = roundUp(16, 16  + 4) = 32

So the minimum buffer size should be 32.
